### ISSUE TITLE
[Donation Proposal]: otlptest

### ISSUE THREAD

**Issue by atoulme** [open-telemetry/community#1880]
### Description

The otlptest project consists of an executable that continuously addresses data to an OTLP endpoint for testing purposes. The project sends in particular invalid histogram data, but can be enhanced to test additional signals or metric scenarios.

### Benefits to the OpenTelemetry community

This would help other implementations of OTLP achieve interoperability with the protocol. In particular, it is useful for other implementations to match edge cases such as too many exemplars, too many buckets, or invalid data situations such as when min is larger than max.

### Reasons for donation

There is no point in keeping this to myself, and this would benefit greatly from additional work on supporting more use cases.

### Repository

https://github.com/atoulme/otlptest

### Existing usage

This is used for testing successfully already.

### Maintenance

I can volunteer as maintainer.

### Licenses

Apache License.

### Trademarks

No trademarks.

**Comment by tigrannajaryan** [comment]
I think this is a useful tool.

We have a [testbed](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/testbed/correctnesstests/traces/correctness_test.go) in the Collector that allows [systematically generating a variety of testdata](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/coreinternal/goldendataset) using PICT framework and sending it to OTLP and other endpoints and also receive it back and compare.

Do you see otlptest as a more standalone tool for testing other endpoints and not just the Collector? I wonder if it is worth extracting the PICT data generator from Collector's testbed and using in this new tool. The PICT generator can probably generate a bigger variety of data than otlptest has.

**Comment by atoulme** [comment]
From my understanding, the testbed tests the collector, while this can be pointed at any backend supporting OTLP to test them out. Maybe there is a way to point testbed to an endpoint, but I haven't seen that.

The tests check edge cases of specific metric points and make minimal assumptions as of yet on the feedback. I would expect folks would run this and perform manual QA to see the impact of ingesting a data point with malformed data.

**Comment by jpkrohling** [comment]
@atoulme, what's the current state of this tool? Has there been progress since the proposal? Should this be part of the Collector, or do you see this as a top-level project?

**Comment by atoulme** [comment]
It's working as intended, it has seen some development since the proposal was made. It has no place with the collector, it is a functional test suite for any OTLP endpoint. It doesn't make assertions on what the behavior should be, but it's also testing limits and fuzzing expectations.

**Comment by atoulme** [comment]
Folks, is there any interest to move this forward in the triaging process?

**Comment by danielgblanco** [comment]
A new repo would require maintainers, etc. Since the collector is the only core component that can expose an OTLP endpoint, would it make sense for it to live under the same contrib codebase?

**Comment by atoulme** [comment]
Yes, that's a good outcome. Please let me know if this is acceptable and I can start discussing with the contrib project.

**Comment by trask** [comment]
hi @atoulme! you don't need to go through the donation proposal process to contribute code to an existing repository

**Comment by atoulme** [comment]
Yes. I am still not sure this is the right idea for the contribution to go to: this component tests the behavior receiving valid and invalid data from a remote endpoint. How the backend reacts to this data is out of scope as of now, with the soft goal of making sure we don't crash it, and is observed outside of the tool.

I will start with a contribution to the opentelemetry-collector-contrib project, but I'm not convinced this is the right place. I will close this issue as I understand that there isn't interest to go in a different direction.

**Comment by atoulme** [comment]
Case in point: I had opened an issue for this in 2023: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28559
